### PR TITLE
DW-2380 add retry policy to DKS interactions

### PR DESCRIPTION
**Pull request by mark-r-m**

### CODE REVIEW

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 1]
```diff
@@ -71,28 +70,49 @@ class HttpKeyServiceTest {
...
+
         keyService.batchDataKey()
+
+        verify(httpClient, times(10)).execute(any(HttpGet::class.java))
+    }
+
+    @Test(expected = DataKeyServiceUnavailableException::class)
+    fun testBatchDataKey_UnknownHttpError_ThrowsException_AndWillRetry() {
+        val statusLine = mock(StatusLine::class.java)
+        //val entity = mock(HttpEntity::class.java)
```
Comment (UdayChokkam) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Can this be removed?

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 2]
```diff
@@ -109,15 +129,17 @@ class HttpKeyServiceTest {
...
         given(httpResponse.entity).willReturn(entity)
         val httpClient = mock(CloseableHttpClient::class.java)
-        given(httpClient.execute(ArgumentMatchers.any(HttpPost::class.java))).willReturn(httpResponse)
+        given(httpClient.execute(any(HttpPost::class.java))).willReturn(httpResponse)
         given(httpClientProvider.client()).willReturn(httpClient)
+
         val dataKeyResult = keyService.decryptKey("123", "ENCRYPTED_KEY_ID")
+
         //val expectedResult: DataKeyResult = Gson().fromJson(responseBody, DataKeyResult::class.java)
         Assert.assertEquals("PLAINTEXT_DATAKEY", dataKeyResult)
```
Comment (UdayChokkam) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: can this be removed as well ?

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 3]
```diff
@@ -134,48 +156,71 @@ class HttpKeyServiceTest {
         given(httpResponse.statusLine).willReturn(statusLine)
         given(httpResponse.entity).willReturn(entity)
         val httpClient = mock(CloseableHttpClient::class.java)
-        given(httpClient.execute(ArgumentMatchers.any(HttpPost::class.java))).willReturn(httpResponse)
+        given(httpClient.execute(any(HttpPost::class.java))).willReturn(httpResponse)
         given(httpClientProvider.client()).willReturn(httpClient)
+
         val dataKeyResult = keyService.decryptKey("123", "ENCRYPTED_KEY_ID")
+
         //val expectedResult: DataKeyResult = Gson().fromJson(responseBody, DataKeyResult::class.java)
```
Comment (UdayChokkam) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: commented code

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 4]
```diff
@@ -134,48 +156,71 @@ class HttpKeyServiceTest {
...
+
+        verify(httpClient, times(10))
+            .execute(any(HttpPost::class.java))
     }
 
     @Autowired
     private lateinit var keyService: HttpKeyService
 
     @Autowired
     private lateinit var httpClientProvider: HttpClientProvider
```
Comment (UdayChokkam) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Looks like more than one line end of the file

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 5]
```diff
@@ -53,7 +52,7 @@ class HttpKeyServiceTest {
     }
 
     @Test
-    fun testDataKeyOk() {
+    fun testBatchDataKey_WillCallClient_AndReturnKey() {
```
Comment (UdayChokkam) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Do we also need tests to check if the delay between retrys is actually 1000ms?

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt (hunk 1 above)
Comment (mark-r-m) [reply src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Done, thanks

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt (hunk 5 above)
Comment (mark-r-m) [reply src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Imho, no, as that is testing Sprint it's self, which is reliable.

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt (hunk 4 above)
Comment (mark-r-m) [reply src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Done, thanks

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt (hunk 3 above)
Comment (mark-r-m) [reply src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt]: Removed, thanks

**Pull request merged**